src/Exchange.js

```diff
@@ -1,33 +1,62 @@
 import { BASE_FACTOR, LIQUIDITY_FEE, TOKENS } from "./constants";
 import { Button, Form } from "react-bootstrap";
-import { encodeToken, tokenToString, formatTokenBalance } from "./helpers";
+import {encodeToken, tokenToString, formatTokenBalance, formatTokenAmount} from "./helpers";
 
 import { ArrowDown } from "react-feather";
 import { ChevronLeft } from "react-feather";
 import { BigInt, add, subtract, multiply, divide } from "jsbi";
 import { find } from "lodash";
-import { default as React, useMemo } from "react";
+import {default as React, useMemo, useState} from "react";
 import { usePostTransaction } from "./mutations";
 import { TokenAmountInput, TokenSelect } from "./Inputs";
 
 export default function Exchange(props) {
   const { onHide, liquidityTokens } = props;
   const [inputAmount, setInputAmount] = React.useState();
+  const [minimumOutputAmount, setMinimumOutputAmount] = React.useState("0");
+  const [exchangeRate, setExchangeRate] = React.useState();
   const [inputToken, setInputToken] = React.useState(TOKENS[0]);
   const [outputToken, setOutputToken] = React.useState(TOKENS[3]);
   const [inputLiquidityToken, setInputLiquidityToken] = React.useState(
     liquidityTokens
   );
+  const [outputLiquidityToken, setOutputLiquidityToken] = React.useState(
+    liquidityTokens
+  );
+  const [error, setError] = React.useState("")
+
   React.useEffect(() => {
-    setInputLiquidityToken(find(liquidityTokens, ["id", inputToken.id]));
+    const token = find(liquidityTokens, ["id", inputToken.id]);
+    setInputLiquidityToken(token);
   }, [inputToken, liquidityTokens]);
+  React.useEffect(() => {
+    const token = find(liquidityTokens, ["id", outputToken.id]);
+    setOutputLiquidityToken(token);
+  }, [outputToken, liquidityTokens]);
+  React.useEffect(() => {
+      let rate = 'N/A';
+      let outputTokenPrice = outputToken.ticker === 'USD' ? 1 : outputLiquidityToken.price || 0;
+      let inputTokenPrice = inputToken.ticker === 'USD' ? 1 : inputLiquidityToken.price || 0;
+      if (outputTokenPrice !== 0) {
+        rate = formatTokenAmount(inputTokenPrice / outputTokenPrice);
+
+      setExchangeRate(`1 ${inputToken.ticker} = ${rate} ${outputToken.ticker}`);
```

Comment:
I've been putting mostly "model stuff" up top of components and "view stuff" down below to separate concerns. Also if a value can be derived from only other values I've been using `useMemo` so:
```javascript
let rate = React.useMemo(() => {
    let outputTokenPrice = outputToken.ticker === 'USD' ? 1 : outputLiquidityToken.price || 0;
    let inputTokenPrice = inputToken.ticker === 'USD' ? 1 : inputLiquidityToken.price || 0;
    outputToken.ticker === 'USD'? formatTokenAmount(inputTokenPrice / outputTokenPrice) : null;
},[outputTokenPrice , inputTokenPrice ])

...
...

<span>{${inputToken.ticker} = ${rate? rate: "N/A"} ${outputToken.ticker}}</span>
```